options-design/DESIGN.md

```diff
@@ -1,52 +1,123 @@
+Each namespace package will have a set of functions that return options for the interfaces contained
+in that package. Each of these functions will return a type that satisfies the interfaces for which
+the option is valid. For example,
+insertopts.BypassDocumentValidation returns a type that satisfies both the insertopts.Many and
+insertopts.One interfaces.
+
+### Bundles
+For each interface in a package there is also a bundle. The purpose of the bundle, as described
+above, is to enable creating a group of options and to provide a discoverable way to find options
+that satisfy a given interface.
+
+Each bundle is named [interface name]Bundle, so the bundle type for the `insertopts.Many` interface
+is named `insertopts.ManyBundle`. For each of the bundle types there is a constructor function
+called Bundle[interface name], with a variadic parameter of the interface and a return of the
+[interface name]Bundle type. For example, the function signature for the `insertopts.BundleMany`
+function would be:
+```
+func BundleMany(...Many) ManyBundle
+```
+
+These bundle types also implement the interface they are associated with. This allows the bundle to
+be passed directly to methods or functions and for bundles to be nested arbitrarily deep. Each
+bundle will have methods attached, each of which mirrors an option function in the package that is a
+valid option for that interface. Finally, each bundle will have an Unbundle method. The purpose of
+this method is to flatten the bundle into the options it was created from. The Unbundle method has a
+deduplicate parameter that will cause the method to only return the last instance of each option
+that was set. The primary purpose behind the Unbundle method is to simplify the usage of these
+option namespaces within the mongo package. The priority of the options is read from left to right,
+with options appearing to the left being overridden by options on the right. This includes initial
+parameters to a bundle, ultimately forming a linked list.
+
+When a user creates a bundle, they'll do something like this:
+```
 BundleFoo(Bar(), Baz(3), Qux(4)).BarBar().BazBaz().Qux(7).Baz(9)
+```
+
+The order those options are processed in is:
+```
+Bar() -> Baz() -> Qux() -> BarBar() -> BazBaz() -> Qux() -> Baz()
+```
 
-The order those options are processed in needs to be Bar() -> Baz() -> Qux() -> BarBar() -> BazBaz()
--> Qux() -> Baz(), with a later thing in the chain overriding the earlier thing in the chain. In
-this example the option values for Qux would be 7 and for Baz would be 9.
+A later option in the chain overrides the earlier option in the chain. In this example the option
+values for Qux would be 7 and for Baz would be 9.
 
-Internally, the mongo package will use Bundles to construct the slice of options that will
-ultimately be used. The way the bundles are designed makes pushes the complexity of unbundling into
+The mongo package will use Bundles to construct the slice of options that will
```

Comment:
Also call out that each CRUD method takes varargs of its associated optioner interface, so that applications are not required to use a bundle if it's not convenient, and applications don't have to pass any options at all in none are needed.

Reply:
This section has nothing to do with the external API, but I'll add an implementation example here to clarify the intent.